### PR TITLE
Fix (Core\Script) Map Calculation Rewrite

### PR DESCRIPTION
**Pull request by acidmanifesto**
## CURRENT STATUS
Only one issue left.
Blink and any front Leap spell seems to clip thru floors and bridges. Terrain ground and gameobjects is fine. Just need to figure out the going thru the floor and bridge issue now.

Purpose of this PR:
This is to address Blink, Warrior's Charge, and other vmap calculated functions that may be problematic in certain cases.

## Changes Proposed:
This is to address various issues of Blink, Warrior's charge, and other vmap related calculated funcitons.

## Issues Addressed:
Errornous vmap calculation handling.

## Tests Performed:
Builds and compiles with no issues. Other scripts adjusted in core with this pr.

## How to Test the Changes:

1. .cheat power on
2. as a priest perform the move blink spell 1953 in any and all terran variations
2. as a warrior perform the move Charge spell 100 in any and all terran variations

## Known Issues and TODO List:

Test any variations where vmap calculations may be needed. 
Test whatever quest\scenario\instance that the spell script spell_svalna_revive_champion is performed. located in northerned map.
Test Exsisting transports to ensure u can still ride from point a to point b (Horde blimps), Flight manager.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by acidmanifesto** [comment]
ok. so extensive testing on my end. there appears to be select issues with blink allowing the priest character to mesh under GAMEOBJECTS. I suspect this is another issues dealing with the version of mmaps we have currently on azerothcore. This PR will need to sit stagnant until we ever upgrade maps to a higher revision.

**Comment by acidmanifesto** [comment]
https://github.com/azerothcore/azerothcore-wotlk/issues/6232

**Comment by acidmanifesto** [comment]
https://github.com/azerothcore/azerothcore-wotlk/issues/7072

**Comment by acidmanifesto** [comment]
Current this is how it looks with this PR.

Show casing Warrior's charge and Blink below:

Warriors Charge:
https://youtu.be/znHdW1R_9zM

Blink with the vmap issues at the end:
https://youtu.be/lOamL5nAveU

So far only blink suffers the issues as we need to have the vmaps be in line with version 4.8

**Comment by claudiodfc** [comment]
Does this need to be tested yet or is this WIP?

**Comment by yuanf225** [comment]
Does this PR now need to be tested or is it in WIP status?

**Comment by acidmanifesto** [comment]
> Does this PR now need to be tested or is it in WIP status?

Until the new map pr passes this pr is going to sit in wip status

**Comment by acidmanifesto** [comment]
https://github.com/azerothcore/azerothcore-wotlk/issues/6232

**Comment by Kitzunu** [comment]
update?

**Comment by acidmanifesto** [comment]
Cant without maps being updated.

**Comment by Kitzunu** [comment]
wdym? Need to re-extract the maps or does this fix not work?

**Comment by acidmanifesto** [comment]
> wdym? Need to re-extract the maps or does this fix not work?

without having newer map revisions, the main issue will become the front leap spells. such as blink. it goes thru gameobjects.
That is really the only thing that is hold this back from further touch up

**Comment by acidmanifesto** [comment]
Currently Status:
Charge does not undermesh
Blink still clips thru gameobjects and structures.

Results:
Need map improvements, which there are other PR's in progress on from other people. will test then when new maps gets deployed.

**Comment by acidmanifesto** [comment]
i need to retest to see if blink is still clipping thru game objects like bridges and stones, houses, whatever.
the thing is we only need my map.cpp and map.h adjustments.
but then we needed the collision to be linked to it via dynamic tree and because of that we ended up with errors in the creature cpp, object cpp and icecrown_citadel.cpp
so one adjust ment literally became a small cluster mess. so i have to look what i need to change in the dynamictree to fix the collison being ignored with blink
then worry about the other files crying error later such as the compile testers.

**Comment by IntelligentQuantum** [comment]
I ported this from TC Maybe it will help for charge idk
https://github.com/IntelligentQuantum/azerothcore-wotlk/commit/e571b09e87ad325c0eec57676234e2397882c3bd

**Comment by acidmanifesto** [comment]
> I ported this from TC Maybe it will help for charge idk [IntelligentQuantum@e571b09](https://github.com/IntelligentQuantum/azerothcore-wotlk/commit/e571b09e87ad325c0eec57676234e2397882c3bd)

i will look into that later. but there was other reasons for this pr besides charge.

**Comment by acidmanifesto** [comment]
## CURRENT STATUS
Only one issue left.
Blink and any front Leap spell seems to clip thru floors and bridges. Terrain ground and gameobjects is fine. Just need to figure out the going thru the floor and bridge issue now.

File: src/server/game/Maps/Map.h [hunk 1]
```diff
@@ -38,6 +38,7 @@
 #include <memory>
 #include <mutex>
 #include <shared_mutex>
+#include <Collision/DynamicTree.h>
```
Comment (Winfidonarleyan) [inline src/server/game/Maps/Map.h:41]:
```suggestion
#include "DynamicTree.h"
```

File: src/server/game/Maps/Map.h [hunk 2]
```diff
@@ -675,7 +676,7 @@ class Map : public GridRefMgr<NGridType>
 
     [[nodiscard]] NGridType* getNGrid(uint32 x, uint32 y) const
     {
-        ASSERT(x < MAX_NUMBER_OF_GRIDS && y < MAX_NUMBER_OF_GRIDS);
+        ASSERT(x < MAX_NUMBER_OF_GRIDS && y < MAX_NUMBER_OF_GRIDS, "x = %u, y = %u", x, y);
```
Comment (Winfidonarleyan) [inline src/server/game/Maps/Map.h:679]:
```suggestion
        ASSERT(x < MAX_NUMBER_OF_GRIDS && y < MAX_NUMBER_OF_GRIDS, "x = {}, y = {}", x, y);
```

File: src/server/game/Maps/Map.cpp [hunk 3]
```diff
@@ -1555,12 +1555,19 @@ uint16 GridMap::getArea(float x, float y) const
...
     if (!_areaMap)
         return _gridArea;
-
-    x = 16 * (32 - x / SIZE_OF_GRIDS);
-    y = 16 * (32 - y / SIZE_OF_GRIDS);
-    int lx = (int)x & 15;
-    int ly = (int)y & 15;
-    return _areaMap[lx * 16 + ly];
+    // Correcting crash in blank lx and ly is not null, what goes out of range because it is not compiled in debug
+    try
```
Comment (UltraNix) [inline src/server/game/Maps/Map.cpp:1559]: ?? Why is this even here?

File: src/common/Collision/DynamicTree.cpp [hunk 4]
```diff
@@ -291,15 +293,16 @@ bool DynamicMapTree::isInLineOfSight(float x1, float y1, float z1, float x2, flo
...
-float DynamicMapTree::getHeight(float x, float y, float z, float maxSearchDist, uint32 phasemask) const
+float DynamicMapTree::getHeight(float x, float y, float z, float maxSearchDist, uint32 phasemask, DynamicTreeCallback* dCallback) const
 {
     G3D::Vector3 v(x, y, z);
     G3D::Ray r(v, G3D::Vector3(0, 0, -1));
     DynamicTreeIntersectionCallback callback(phasemask, VMAP::ModelIgnoreFlags::Nothing);
     impl->intersectZAllignedRay(r, callback, maxSearchDist);
 
-    if (callback.didHit())
+    if (dCallback)
```
Comment (UltraNix) [inline src/common/Collision/DynamicTree.cpp:303]: `dCallback` is always null, so this is completely useless.

File: src/server/game/Maps/Map.cpp [hunk 5]
```diff
@@ -2478,12 +2485,20 @@ bool Map::GetObjectHitPos(uint32 phasemask, float x1, float y1, float z1, float
...
+float Map::GetHeight(uint32 phasemask, float x, float y, float z, bool vmap /*= true*/, float maxSearchDist /*= DEFAULT_HEIGHT_SEARCH*/) const
 {
-    float h1, h2;
-    h1 = GetHeight(x, y, z, vmap, maxSearchDist);
-    h2 = _dynamicTree.getHeight(x, y, z, maxSearchDist, phasemask);
-    return std::max<float>(h1, h2);
+    DynamicTreeCallback* dCallback = nullptr;
+    float vmapZ = GetHeight(x, y, z, vmap, maxSearchDist);
+    float goZ   = _dynamicTree.getHeight(x, y, z, maxSearchDist, phasemask, dCallback);
+    if (vmapZ > goZ && dCallback)
```
Comment (UltraNix) [inline src/server/game/Maps/Map.cpp:2493]: `dCallback` is always null - what this changes?

File: src/server/game/Maps/Map.cpp (hunk 3 above)
Comment (acidmanifesto) [reply src/server/game/Maps/Map.cpp:1559]: Its a basic try and catch.
The try statement allows you to define a block of code to be tested for errors while it is being executed.
The catch statement allows you to define a block of code to be executed, if an error occurs in the try block.
Try and catch is always in pairs.

File: src/server/game/Maps/Map.cpp (hunk 5 above)
Comment (acidmanifesto) [reply src/server/game/Maps/Map.cpp:2493]: Stops the undermeshing.

File: src/server/game/Maps/Map.cpp (hunk 3 above)
Comment (UltraNix) [reply src/server/game/Maps/Map.cpp:1559]: Yes, I know what is the purpose. I don't know why it is here. What errors could it be here?

Comment (acidmanifesto) [reply src/server/game/Maps/Map.cpp:1559]: Out of range values leads to you falling thru the terrain.

**Comment by UltraNix** [comment]
Sorry, but to me all the changes done here are nonsenses

**Comment by acidmanifesto** [comment]
> Sorry, but to me all the changes done here are nonsenses

it stops the undermshing and been stopping all undermeshing for a while. i encourage u to try the pr and see first hand.

**Comment by UltraNix** [comment]
I don't need to test it - I just see the code. `Map::GetHeight` always returns `VMAP_INVALID_HEIGHT_VALUE` so broke many things.
E.g.
- .gps always returns invalid ground and floor z
- non-flyable mobs can fly to you (just `.gm fly`, ascend and `.cometome` to any mob
- any of the gameobjects with collision (elevators, ships, etc) get wrong ground position
- many spells with destination are also broken, e.g. blink, 
- flying mobs on death gets wrong fall position
- and many more...

**Comment by acidmanifesto** [comment]
> I don't need to test it - I just see the code. `Map::GetHeight` always returns `VMAP_INVALID_HEIGHT_VALUE` so broke many things. E.g.
> 
> * .gps always returns invalid ground and floor z
> * non-flyable mobs can fly to you (just `.gm fly`, ascend and `.cometome` to any mob
> * any of the gameobjects with collision (elevators, ships, etc) get wrong ground position
> * many spells with destination are also broken, e.g. blink, fishing
> * flying mobs on death gets wrong fall position
> * confused movement generator (e.g. blind) completely broken
> * and many more...

if you feel like all this is happening with out testing then you are welcome to close the PR.

**Comment by UltraNix** [comment]
Yes - I'm pretty sure.

**Pull request closed without merging**